File: Grind75/accountsMerge.py
from typing import List, Dict

def accountsMerge(accounts: List[List[str]]) -> List[List[str]]:
    """
    Merge accounts by connecting emails that appear in the same account.
    Returns a list where each element is [name, sorted_email_1, sorted_email_2, ...].
    """
    # Union-Find data structure for emails
    parent: Dict[str, str] = {}

    def find(x: str) -> str:
        # Path compression
        if parent[x] != x:
            parent[x] = find(parent[x])
        return parent[x]

    def union(a: str, b: str) -> None:
        root_a = find(a)
        root_b = find(b)
        if root_a != root_b:
            parent[root_b] = root_a

    email_to_name: Dict[str, str] = {}

    # Step 1: Initialize parent pointers and union emails within each account
    for account in accounts:
        name = account[0]
        first_email = account[1]

        for email in account[1:]:
            # Initialize parent pointer if seeing this email for the first time
            if email not in parent:
                parent[email] = email
            # Record the name for this email
            email_to_name[email] = name
            # Union the first email in this account with the current one
            union(first_email, email)

    # Step 2: Group emails by their root parent
    groups: Dict[str, List[str]] = {}
    for email in parent:
        root = find(email)
        groups.setdefault(root, []).append(email)

    # Step 3: Build the merged account lists
    merged: List[List[str]] = []
    for root, emails in groups.items():
        # Sort the emails for this group
        emails.sort()
        # The name is the same for all emails in this group
        name = email_to_name[root]
        merged.append([name] + emails)

    return merged


# Time Complexity: O(n α(n) + n log n) ≈ O(n log n),
#   where n = total number of emails across all accounts,
#   α(n) is the inverse-Ackermann function from Union-Find (effectively constant),
#   and n log n comes from sorting each group of emails.
# Space Complexity: O(n),
#   for storing parent pointers, the email-to-name map, and the groups of emails.

File: Grind75/accountsMerge.js
class UnionFind {
  constructor() {
    this.parent = new Map();
  }

  find(x) {
    if (!this.parent.has(x)) {
      this.parent.set(x, x);
    }
    if (this.parent.get(x) !== x) {
      this.parent.set(x, this.find(this.parent.get(x))); // path compression
    }
    return this.parent.get(x);
  }

  union(x, y) {
    const rootX = this.find(x);
    const rootY = this.find(y);
    if (rootX !== rootY) {
      this.parent.set(rootX, rootY);
    }
  }
}

function accountsMerge(accounts) {
  const uf = new UnionFind();
  const emailToName = new Map();

  for (const [name, ...emails] of accounts) {
    for (const email of emails) {
      emailToName.set(email, name);
      uf.find(email); // initialize in union-find
    }

    for (let i = 1; i < emails.length; i++) {
      uf.union(emails[0], emails[i]);
    }
  }

  // Group emails by their root parent
  const rootToEmails = new Map();
  for (const email of emailToName.keys()) {
    const root = uf.find(email);
    if (!rootToEmails.has(root)) {
      rootToEmails.set(root, []);
    }
    rootToEmails.get(root).push(email);
  }

  // Format result
  const result = [];
  for (const [root, emails] of rootToEmails.entries()) {
    emails.sort();
    result.push([emailToName.get(root), ...emails]);
  }

  return result;
}
